sw/device/lib/dif/dif_kmac.c

```diff
@@ -4,4 +4,478 @@
+  uint32_t cmd_reg =
+      bitfield_field32_write(0, KMAC_CMD_CMD_FIELD, KMAC_CMD_CMD_VALUE_START);
+  mmio_region_write32(kmac->params.base_addr, KMAC_CMD_REG_OFFSET, cmd_reg);
+
+  // Poll until the status register is in the 'absorb' state.
+  while (true) {
+    if (is_state_absorb(kmac->params)) {
+      break;
+    }
+    // TODO(#6248): check for error.
+  }
+
+  return kDifKmacOk;
+}
+
+dif_kmac_result_t dif_kmac_absorb(dif_kmac_t *kmac, const void *msg, size_t len,
+                                  size_t *processed) {
+  // Set the number of bytes processed to 0.
+  if (processed != NULL) {
+    *processed = 0;
+  }
+
+  if (kmac == NULL || (msg == NULL && len != 0)) {
+    return kDifKmacBadArg;
+  }
+
+  // Check that an operation has been started.
+  if (kmac->r == 0) {
+    return kDifKmacError;
+  }
+
+  // Poll until the the status register is in the 'absorb' state.
+  if (!is_state_absorb(kmac->params)) {
+    return kDifKmacError;
+  }
+
+  // Copy the message one byte at a time.
+  // This could be sped up copying a word at a time but be careful
+  // about message endianness (e.g. only copy a word at a time when in
+  // little-endian mode).
+  for (size_t i = 0; i < len; ++i) {
+    mmio_region_write8(kmac->params.base_addr, KMAC_MSG_FIFO_REG_OFFSET,
+                       ((const uint8_t *)msg)[i]);
+  }
+
+  if (processed != NULL) {
+    *processed = len;
+  }
+  return kDifKmacOk;
+}
+
+dif_kmac_result_t dif_kmac_squeeze(dif_kmac_t *kmac, void *out, size_t len,
+                                   size_t *processed) {
+  if (kmac == NULL || (out == NULL && len != 0)) {
+    return kDifKmacBadArg;
+  }
+
+  // Set `processed` to 0 so we can return early without setting it again.
+  if (processed != NULL) {
+    *processed = 0;
+  }
+
+  // Move into squeezing state (if not already in it).
+  // Do this even if the length requested is 0 or too big.
+  if (!kmac->squeezing) {
+    kmac->squeezing = true;
+
+    // Issue squeeze command.
+    uint32_t cmd_reg = bitfield_field32_write(0, KMAC_CMD_CMD_FIELD,
+                                              KMAC_CMD_CMD_VALUE_PROCESS);
+    mmio_region_write32(kmac->params.base_addr, KMAC_CMD_REG_OFFSET, cmd_reg);
+  }
+
+  // If the operation has a fixed length output then the total number of bytes
+  // requested must not exceed that length.
+  if (kmac->d != 0 && len > (kmac->d - kmac->offset)) {
+    return kDifKmacError;
+  }
+
+  if (len == 0) {
+    return kDifKmacOk;
+  }
+
+  // Poll the status register until in the 'squeeze' state.
+  while (true) {
+    if (is_state_squeeze(kmac->params)) {
+      break;
+    }
+    // TODO(#6248): check for error.
+  }
+
+  // Read out the state shares.
+  while (len > 0) {
+    size_t n = len;
+    size_t remaining = (kmac->d == 0 ? kmac->r : kmac->d) - kmac->offset;
+    if (n > remaining) {
+      n = remaining;
+    }
+    if (n == 0) {
+      // TODO(XOF): request more state.
+      return kDifKmacError;
+    }
+
```

Comment:
The following code to copy the state into the user provided buffer could be done in a smarter way with fewer copies. I think this is OK for now though. I'll revisit it as I implement the XOF functions and add more testing for multiple squeeze calls.